src/ParallelAlgorithms/Events/MonitorMemory.hpp

```diff
@@ -0,0 +1,351 @@
+        auto array_element_proxy =
+            Parallel::get_parallel_component<component>(cache)[array_index];
+        const double size_in_bytes = static_cast<double>(
+            size_of_object_in_bytes(*Parallel::local(array_element_proxy)));
+        const double size_in_megabytes = size_in_bytes / 1.0e6;
+
+        // vector the size of the number of nodes we are running on. Set the
+        // 'my_node'th element of the vector to the size of this Element. Then
+        // when we reduce, we will have a vector with 'num_nodes' elements, each
+        // of which represents the total memory usage of all Elements on that
+        // node.
+        const size_t num_nodes = Parallel::number_of_nodes<size_t>(
+            *Parallel::local(array_element_proxy));
+        const size_t my_node =
+            Parallel::my_node<size_t>(*Parallel::local(array_element_proxy));
+        std::vector<double> data(num_nodes, 0.0);
+        data[my_node] = size_in_megabytes;
+
+        Parallel::contribute_to_reduction<
+            mem_monitor::ProcessArray<ParallelComponent>>(
+            ReductionData{static_cast<double>(observation_value), data},
+            array_element_proxy, memory_monitor_proxy);
+      } else if constexpr (Parallel::is_singleton_v<component>) {
+        // If this is a singleton, we only run this once so use the designated
+        // element. Nothing to reduce with singletons so just call the simple
+        // action on the singleton
+        if (designated_element) {
+          auto& singleton_proxy =
+              Parallel::get_parallel_component<component>(cache);
+
+          Parallel::simple_action<mem_monitor::ProcessSingleton>(
+              singleton_proxy, static_cast<double>(observation_value));
+        }
+      } else if constexpr (Parallel::is_nodegroup_v<component> or
+                           Parallel::is_group_v<component>) {
+        // If this is a (node)group, call a simple action on each branch if on
+        // the designated element
+        if (designated_element) {
+          // Can't run simple actions on the cache so broadcast a specific entry
+          // method that will calculate the size and send it to the memory
+          // monitor
+          if constexpr (std::is_same_v<component,
+                                       Parallel::GlobalCache<Metavariables>>) {
+            auto cache_proxy = cache.get_this_proxy();
+
+            // This will be called on all branches of the GlobalCache
+            cache_proxy.compute_size_for_memory_monitor(
+                static_cast<double>(observation_value));
+          } else if constexpr (std::is_same_v<component,
+                                              Parallel::MutableGlobalCache<
+                                                  Metavariables>>) {
+            // Even though we can't monitor the MutableGlobalCache at the
+            // moment, we still add it in for consistency.
```

Comment:
I'm confused by what you're asking. This constexpr branch for the MutableGlobalCache can never be reached because I removed the MutableGlobalCache from the component list above.

Reply:
Oh, OK, I see what it is doing, now.  I didn't read this as a "can't get here" comment.  I interpreted the "it" in this sentence to refer to the output for the MutableGlobalCache, rather than the code branch.

Reply:
I reworded the comment to make it clearer